### PR TITLE
vtep: VTEP redirection in bpf_host  from-host  to support  L7 policy

### PR DESCRIPTION
**Pull request by vincentmli**
L7 policies redirect packets on egress of bpf_lxc to userspace
    proxy. When packets leave the proxy, they don't go back through
    bpf_lxc again, so the VTEP redirectons won't be enforced. Add
    the VTEP redirection in bpf_host. This also requires route to
    VTEP CIDR setup like:
    
    <VTEP CIDR> dev cilium_host scope link
    
    Reason is when packets are passed to host stack with destination
    ip in VTEP CIDR, host stack needs to know where to route the
    packets to, and __section("from-host") is attached to cilium_host,
    so we setup route on cilium_host with scope link.
 
    we still keep the VTEP redirection in bpf_lxc when L7 proxy is disabled
    VTEP redirection would happen earlier to avoid packets passing to
    host stack to be handled in bpf_host.

Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

Fixes: #issue-number

```release-note
Add support for L7 policies with VTEP integration
```

### CODE REVIEW

**Comment by vincentmli** [comment]
Hi @pchaigno @joestringer, this is the PR to address L7 policy support, please let me know if this is the right approach, especially if the route is really needed. here is my reasoning

for example my vtep cidr is 10.1.1.0/24, pod cidr is 10.0.0.0/16, when pod send packet to 10.1.1.1/32 within vtep cidr, after L7 policy processing, the packet is passed to host stack with target ip 10.1.1.1/32, there could be default route in host to handle node traffic or management network traffic, but 10.1.1.1/32 goes nowhere since it is not visible to the node network or management network, 10.1.1.1/32 has to be routed/directed to cilium_vxlan for VXLAN encap,  a route to vtep cidr via cilium_host scope seems does the work.

File: daemon/cmd/datapath.go [hunk 1]
```diff
@@ -275,6 +279,12 @@ func (d *Daemon) syncEndpointsAndHostIPs() error {
 		if err != nil {
 			return err
 		}
+		if option.Config.EnableL7Proxy {
+			err = setupRouteToVtepCidr()
```
Comment (joestringer) [inline daemon/cmd/datapath.go]: Minor nit here, only applying the route diff if the L7 proxy is enabled means that if you disable L7 proxy... the host routes will stay populated. However I think you intended that if L7 proxy is disabled, then all routes are removed? So probably this condition should wrap just the "addevVtepRoutes" case inside this function.

File: bpf/bpf_lxc.c [hunk 2]
```diff
@@ -895,7 +895,11 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
 skip_egress_gateway:
 #endif
 
-#ifdef ENABLE_VTEP
+	/* L7 proxy result in VTEP redirection in bpf_host, but when L7 proxy disabled
+	 * We want VTEP redirection handled earlier here to avoid packets passing to
+	 * stack to bpf_host for VTEP redirection.
+	 */
+#if defined(ENABLE_VTEP) && !defined(ENABLE_L7_PROXY)
```
Comment (joestringer) [inline bpf/bpf_lxc.c]: I'm a bit confused, isn't this the one path that guarantees that traffic from a pod will get routed to the VTEP correctly? Doesn't this change break VTEP support if you have enabled L7 proxy?

**Review by joestringer — CHANGES_REQUESTED** [review]
Seems like a reasonable approach at a glance, just a couple of comments on the details of how the logic is enabled or disabled based on the L7 proxy flag.

File: daemon/cmd/datapath.go (hunk 1 above)
Comment (vincentmli) [reply daemon/cmd/datapath.go]: make perfect sense, I have not thought about that

File: bpf/bpf_lxc.c (hunk 2 above)
Comment (vincentmli) [reply bpf/bpf_lxc.c]: it does not break VTEP support, `!defined(ENABLE_L7_PROXY)` was added when I was testing the VTEP redirection in bpf_host to make sure VTEP redirection works in bpf_host, Paul and I also noticed that when L7 proxy enabled, but no L7 policy applied to pod, the VTEP redirection actually happens in bpf_lxc. I do prefer your choice though to remove `!defined(ENABLE_L7_PROXY)` to guarantees the traffic from pod will get routed to VTEP when L7 proxy disabled, or when L7 proxy enabled but no L7 policy applied to pod.

File: daemon/cmd/datapath.go [hunk 3]
```diff
@@ -514,6 +517,124 @@ func setupVTEPMapping() error {
...
+				Priority: linux_defaults.RulePriorityVtep,
+				To:       prefix.IPNet,
+				Table:    linux_defaults.RouteTableVtep,
+			}
+			if err := route.DeleteRule(rule); err != nil {
+				return fmt.Errorf("Delete VTEP CIDR rule error: %w", err)
+			}
+		}
+	} else {
+		// remove all VTEP routes and rules if L7 proxy disabled
```
Comment (joestringer) [inline daemon/cmd/datapath.go]: This code seems like it's duplicated with the same code inside the `option.Config.EnableL7Proxy` case above, is there a reason to have them separate? Why not just have the `if option.Config.EnableL7Proxy  { ... }` case only cover the add routes logic?

Comment (vincentmli) [reply daemon/cmd/datapath.go]: option.Config.EnableL7Proxy is to serve 1) add new VTEP cidr route, 2) remove old VTEP cidr route if user changed VTEP cidr like `cilium config set vtep-cdir "new vtep cidr"`

**Comment by vincentmli** [comment]
did rebase on master branch, cilium agent fail to get up and running with error logs

```
grep -w 'error=' /tmp/ca-log.sh
level=info msg="Auto-disabling \"enable-bpf-clock-probe\" feature since KERNEL_HZ cannot be determined" error="Cannot probe CONFIG_HZ" subsys=daemon
level=debug msg="Controller run failed" consecutiveErrors=1 error="CiliumNode cache store not yet initialized" name="sync-node-with-ciliumnode (cilium-dev)" subsys=controller uuid=32e3daa2-1005-4d75-85f2-6b0396555490
level=debug msg="Controller run failed" consecutiveErrors=1 error="failed to inject labels into ipcache: k8s cache not fully synced" name=ipcache-inject-labels subsys=controller uuid=33c75767-4e6f-45aa-bf57-db7649b066cb
level=warning msg="Failed to remove old router IPs (restored IP: 10.0.0.167) from cilium_host. Manual intervention is required to remove all other old IPs." error="Link not found" subsys=daemon
level=debug msg="Controller run failed" consecutiveErrors=2 error="failed to inject labels into ipcache: k8s cache not fully synced" name=ipcache-inject-labels subsys=controller uuid=33c75767-4e6f-45aa-bf57-db7649b066cb
level=debug msg="Controller run failed" consecutiveErrors=3 error="failed to inject labels into ipcache: k8s cache not fully synced" name=ipcache-inject-labels subsys=controller uuid=33c75767-4e6f-45aa-bf57-db7649b066cb
level=debug msg="Group not found" error="group: unknown group cilium" file-path=/var/run/cilium/monitor1_2.sock group=cilium subsys=api
level=debug msg="Couldn't open CT map for upgrade" error="Unable to get object /sys/fs/bpf/tc/globals/cilium_ct6_global: no such file or directory" file-path=/sys/fs/bpf/tc/globals/cilium_ct6_global subsys=map-ct
level=debug msg="Couldn't open CT map for upgrade" error="Unable to get object /sys/fs/bpf/tc/globals/cilium_ct_any6_global: no such file or directory" file-path=/sys/fs/bpf/tc/globals/cilium_ct_any6_global subsys=map-ct
level=debug msg="Cannot establish connection to local cilium instance" error="Get \"http://%2Fvar%2Frun%2Fcilium%2Fcilium.sock/v1/healthz\": dial unix /var/run/cilium/cilium.sock: connect: no such file or directory" subsys=cilium-health-launcher
level=debug msg="Didn't find existing device" error="Link not found" subsys=cilium-health-launcher veth=cilium_health
level=debug msg="Didn't find existing device" error="Link not found" subsys=cilium-health-launcher veth=lxc_health
level=debug msg="Unable to remove cilium-health namespace" error="Unable to delete named netns cilium-health: Cannot remove namespace file \"/var/run/netns/cilium-health\": No such file or directory\n exit status 1" subsys=cilium-health-launcher
level=debug msg="Group not found" error="group: unknown group cilium" file-path=/var/run/cilium/cilium.sock group=cilium subsys=api
level=debug msg="Group not found" error="group: unknown group cilium" file-path=/var/run/cilium/hubble.sock group=cilium subsys=api
level=debug msg="API call has been processed" error="endpoint not found" name=endpoint-delete processingDuration="2.064µs" subsys=rate totalDuration="62.978µs" uuid=04e5a28c-1e3a-4779-83c8-3892fbd83dbd waitDurationTotal="51.286µs"
level=debug msg="API call has been processed" error="endpoint not found" name=endpoint-delete processingDuration="1.703µs" subsys=rate totalDuration="42.589µs" uuid=d4a14e2d-4e13-48a0-bc5c-c9ff3defd350 waitDurationTotal="32.59µs"
level=debug msg="API call has been processed" error="endpoint not found" name=endpoint-delete processingDuration="3.646µs" subsys=rate totalDuration="61.345µs" uuid=5a4a1482-2548-4d91-a6a9-269bbc031cbf waitDurationTotal="40.586µs"
level=debug msg="API call has been processed" error="endpoint not found" name=endpoint-delete processingDuration="2.104µs" subsys=rate totalDuration="41.738µs" uuid=13872bea-cb85-4e36-861c-98085e108f67 waitDurationTotal="30.838µs"
level=debug msg="API call has been processed" error="endpoint not found" name=endpoint-delete processingDuration="2.795µs" subsys=rate totalDuration="106.199µs" uuid=48b9e6b5-cca7-4151-af68-f681edf874ba waitDurationTotal="87.624µs"
level=debug msg="API call has been processed" error="endpoint not found" name=endpoint-delete processingDuration="2.264µs" subsys=rate totalDuration="46.166µs" uuid=75c413ad-295d-4897-9a28-d8043cbd83d3 waitDurationTotal="35.416µs"
level=debug msg="API call has been processed" error="endpoint not found" name=endpoint-delete processingDuration="3.045µs" subsys=rate totalDuration="56.776µs" uuid=e25e10c5-c60f-41e3-b098-e44b3cc18cc7 waitDurationTotal="42.94µs"
level=debug msg="API call has been processed" error="endpoint not found" name=endpoint-delete processingDuration="2.675µs" subsys=rate totalDuration="50.044µs" uuid=fe618142-efab-4d07-bf93-3b5c76547e9b waitDurationTotal="33.443µs"
level=debug msg="Cannot find socket" error="stat /var/run/cilium/health.sock: no such file or directory" file-path=/var/run/cilium/health.sock subsys=cilium-health-launcher
level=debug msg="Cannot find socket" error="stat /var/run/cilium/health.sock: no such file or directory" file-path=/var/run/cilium/health.sock subsys=cilium-health-launcher
level=debug msg="Greeting failed" error="Get \"http://10.169.72.233:4240/hello\": dial tcp 10.169.72.233:4240: connect: connection refused" host="http://10.169.72.233:4240" ipAddr=10.169.72.233 nodeName=centos-dev.localdomain path="Via L3" subsys=health-server

..........
level=debug msg="Cannot find socket" error="stat /var/run/cilium/health.sock: no such file or directory" file-path=/var/run/cilium/health.sock subsys=cilium-health-launcher
level=debug msg="Finished writing ELF" error="<nil>" new-elf-path=3830_next/bpf_lxc.o subsys=elf template-path=/var/run/cilium/state/templates/68c8901056d0da43a2550ec0ef2e91c4d19e37bbc3d2655163c59cc0ba558b42/bpf_lxc.o
level=debug msg="Finished writing ELF" error="<nil>" new-elf-path=1859_next/bpf_lxc.o subsys=elf template-path=/var/run/cilium/state/templates/68c8901056d0da43a2550ec0ef2e91c4d19e37bbc3d2655163c59cc0ba558b42/bpf_lxc.o
level=debug msg="Finished writing ELF" error="<nil>" new-elf-path=2474_next/bpf_lxc.o subsys=elf template-path=/var/run/cilium/state/templates/68c8901056d0da43a2550ec0ef2e91c4d19e37bbc3d2655163c59cc0ba558b42/bpf_lxc.o
level=debug msg="Finished writing ELF" error="<nil>" new-elf-path=696_next/bpf_lxc.o subsys=elf template-path=/var/run/cilium/state/templates/68c8901056d0da43a2550ec0ef2e91c4d19e37bbc3d2655163c59cc0ba558b42/bpf_lxc.o
level=debug msg="Finished writing ELF" error="<nil>" new-elf-path=555_next/bpf_lxc.o subsys=elf template-path=/var/run/cilium/state/templates/68c8901056d0da43a2550ec0ef2e91c4d19e37bbc3d2655163c59cc0ba558b42/bpf_lxc.o
level=debug msg="Finished writing ELF" error="<nil>" new-elf-path=3289_next/bpf_host.o subsys=elf template-path=/var/run/cilium/state/templates/c4a09b034e1867d4fd051f931859e044c5d2d9bb5c7dcc10d20273966852f3b7/bpf_host.o
level=debug msg="Finished writing ELF" error="<nil>" new-elf-path=3289_next/bpf_host_cilium_net.o subsys=elf template-path=3289_next/bpf_host.o
level=debug msg="Finished writing ELF" error="<nil>" new-elf-path=3289_next/bpf_netdev_ens192.o subsys=elf template-path=3289_next/bpf_host.o
level=debug msg="Finished writing ELF" error="<nil>" new-elf-path=3289_next/bpf_netdev_ens224.o subsys=elf template-path=3289_next/bpf_host.o
...........
level=debug msg="Cannot find socket" error="stat /var/run/cilium/health.sock: no such file or directory" file-path=/var/run/cilium/health.sock subsys=cilium-health-launcher
level=debug msg="Group not found" error="group: unknown group cilium" file-path=/var/run/cilium/health.sock group=cilium subsys=api

```

File: daemon/cmd/datapath.go (hunk 3 above)
Comment (joestringer) [reply daemon/cmd/datapath.go]: Does the code run the same logic in (2) from your explanation from above and also this else condition?

**Comment by vincentmli** [comment]
I usually use helm template to create yaml file and kubectl to apply the yaml file, I did not use helm template after master rebase, let me try that.

File: daemon/cmd/datapath.go (hunk 3 above)
Comment (vincentmli) [reply daemon/cmd/datapath.go]: the else condition is when L7 proxy disabled, remove all existing VTEP routes/rules in the node, for scenario that user enabled L7 proxy, but later disabled L7 proxy.  2) is L7 proxy enabled, but user want to change the vtep cidr from old vtep cidr to new vtep cidr, in this case, we need to remove the old vtep cidr route/rules. so else and 2) code logic is different. hope I answered your question :)

Comment (joestringer) [reply daemon/cmd/datapath.go]: Ah I see thanks for the explanation. `for _, prefix := range routeCidrs {` is different.

I think we could avoid about 30 lines of duplicate code by applying this diff on top (basically just set the `removedVtepRoutes` to the right slice if L7 is disabled, then un-indent the main removedVtepRoutes code so it's the same between the (2) and else cases):

```
diff --git a/daemon/cmd/datapath.go b/daemon/cmd/datapath.go
index 3a49dd5cf7e5..abcc8ba733a6 100644
--- a/daemon/cmd/datapath.go
+++ b/daemon/cmd/datapath.go
@@ -571,64 +571,36 @@ func setupRouteToVtepCidr() error {
                                return fmt.Errorf("Update VTEP CIDR rule error: %w", err)
                        }
                }
+       } else {
+               removedVtepRoutes = routeCidrs
+       }
 
-               for _, prefix := range removedVtepRoutes {
-                       ip4 := prefix.IP.To4()
-                       if ip4 == nil {
-                               return fmt.Errorf("Invalid VTEP CIDR IPv4 address: %v", ip4)
-                       }
-                       r := route.Route{
-                               Device: defaults.HostDevice,
-                               Prefix: *prefix.IPNet,
-                               Scope:  netlink.SCOPE_LINK,
-                               MTU:    vtepMTU,
-                               Table:  linux_defaults.RouteTableVtep,
-                       }
-                       if err := route.Delete(r); err != nil {
-                               return fmt.Errorf("Delete VTEP CIDR route error: %w", err)
-                       }
-                       log.WithFields(logrus.Fields{
-                               logfields.IPAddr: r.Prefix.String(),
-                       }).Info("VTEP route removed")
-
-                       rule := route.Rule{
-                               Priority: linux_defaults.RulePriorityVtep,
-                               To:       prefix.IPNet,
-                               Table:    linux_defaults.RouteTableVtep,
-                       }
-                       if err := route.DeleteRule(rule); err != nil {
-                               return fmt.Errorf("Delete VTEP CIDR rule error: %w", err)
-                       }
+       for _, prefix := range removedVtepRoutes {
+               ip4 := prefix.IP.To4()
+               if ip4 == nil {
+                       return fmt.Errorf("Invalid VTEP CIDR IPv4 address: %v", ip4)
                }
-       } else {
-               // remove all VTEP routes and rules if L7 proxy disabled
-               for _, prefix := range routeCidrs {
-                       ip4 := prefix.IP.To4()
-                       if ip4 == nil {
-                               return fmt.Errorf("Invalid VTEP CIDR IPv4 address: %v", ip4)
-                       }
-                       r := route.Route{
-                               Device: defaults.HostDevice,
-                               Prefix: *prefix.IPNet,
-                               Scope:  netlink.SCOPE_LINK,
-                               MTU:    vtepMTU,
-                               Table:  linux_defaults.RouteTableVtep,
-                       }
-                       if err := route.Delete(r); err != nil {
-                               return fmt.Errorf("Delete VTEP CIDR route error: %w", err)
-                       }
-                       log.WithFields(logrus.Fields{
-                               logfields.IPAddr: r.Prefix.String(),
-                       }).Info("VTEP route removed")
+               r := route.Route{
+                       Device: defaults.HostDevice,
+                       Prefix: *prefix.IPNet,
+                       Scope:  netlink.SCOPE_LINK,
+                       MTU:    vtepMTU,
+                       Table:  linux_defaults.RouteTableVtep,
+               }
+               if err := route.Delete(r); err != nil {
+                       return fmt.Errorf("Delete VTEP CIDR route error: %w", err)
+               }
+               log.WithFields(logrus.Fields{
+                       logfields.IPAddr: r.Prefix.String(),
+               }).Info("VTEP route removed")
 
-                       rule := route.Rule{
-                               Priority: linux_defaults.RulePriorityVtep,
-                               To:       prefix.IPNet,
-                               Table:    linux_defaults.RouteTableVtep,
-                       }
-                       if err := route.DeleteRule(rule); err != nil {
-                               return fmt.Errorf("Delete VTEP CIDR rule error: %w", err)
-                       }
+               rule := route.Rule{
+                       Priority: linux_defaults.RulePriorityVtep,
+                       To:       prefix.IPNet,
+                       Table:    linux_defaults.RouteTableVtep,
+               }
+               if err := route.DeleteRule(rule); err != nil {
+                       return fmt.Errorf("Delete VTEP CIDR rule error: %w", err)
                }
        }
 
```

Comment (vincentmli) [reply daemon/cmd/datapath.go]: ah, that is good code suggestion :)

**Review by joestringer — APPROVED** [review]
LGTM thanks. Were you able to test this, for example by configuring this in an environment with the VTEP providing world access, then running the `cilium connectivity check`? This should demonstrate L7 functionality via the DNS+FQDN policies.

**Comment by vincentmli** [comment]
after  I enable `enable-envoy-config: "true"` after master rebase and redo helm template, I got error:

```
level=error msg=k8sError error="github.com/cilium/cilium/pkg/k8s/watchers/cilium_clusterwide_envoy_config.go:81: Failed to watch *v2.CiliumClusterwideEnvoyConfig: failed to list *v2.CiliumClusterwideEnvoyConfig: the server could not find the requested resource (get ciliumclusterwideenvoyconfigs.cilium.io)" subsys=k8s
```
no such error before master rebase. no problem if I set `enable-envoy-config: "false"`

**Comment by vincentmli** [comment]
> LGTM thanks. Were you able to test this, for example by configuring this in an environment with the VTEP providing world access, then running the `cilium connectivity check`? This should demonstrate L7 functionality via the DNS+FQDN policies.

yes i tested applying L7 egress policy https://github.com/cilium/cilium-service-mesh-beta/tree/main/l7-traffic-management to pod and send traffic to VTEP, it works, I can see the envoy server header. but now I have problem get cilium agent up and running after `enable-envoy-config: "true"` after master rebase and redo helm template, agent errors out like:
```
level=error msg=k8sError error="github.com/cilium/cilium/pkg/k8s/watchers/cilium_envoy_config.go:82: Failed to watch *v2.CiliumEnvoyConfig: failed to list *v2.CiliumEnvoyConfig: the server could not find the requested resource (get ciliumenvoyconfigs.cilium.io)" subsys=k8s
level=warning msg="github.com/cilium/cilium/pkg/k8s/watchers/cilium_clusterwide_envoy_config.go:81: failed to list *v2.CiliumClusterwideEnvoyConfig: the server could not find the requested resource (get ciliumclusterwideenvoyconfigs.cilium.io)" subsys=klog
level=error msg=k8sError error="github.com/cilium/cilium/pkg/k8s/watchers/cilium_clusterwide_envoy_config.go:81: Failed to watch *v2.CiliumClusterwideEnvoyConfig: failed to list *v2.CiliumClusterwideEnvoyConfig: the server could not find the requested resource (get ciliumclusterwideenvoyconfigs.cilium.io)" subsys=k8s
level=warning msg="github.com/cilium/cilium/pkg/k8s/watchers/cilium_clusterwide_envoy_config.go:81: failed to list *v2.CiliumClusterwideEnvoyConfig: the server could not find the requested resource (get ciliumclusterwideenvoyconfigs.cilium.io)" subsys=klog
level=error msg=k8sError error="github.com/cilium/cilium/pkg/k8s/watchers/cilium_clusterwide_envoy_config.go:81: Failed to watch *v2.CiliumClusterwideEnvoyConfig: failed to list *v2.CiliumClusterwideEnvoyConfig: the server could not find the requested resource (get ciliumclusterwideenvoyconfigs.cilium.io)" subsys=k8s
level=warning msg="github.com/cilium/cilium/pkg/k8s/watchers/cilium_envoy_config.go:82: failed to list *v2.CiliumEnvoyConfig: the server could not find the requested resource (get ciliumenvoyconfigs.cilium.io)" subsys=klog
level=error msg=k8sError error="github.com/cilium/cilium/pkg/k8s/watchers/cilium_envoy_config.go:82: Failed to watch *v2.CiliumEnvoyConfig: failed to list *v2.CiliumEnvoyConfig: the server could not find the requested resource (get ciliumenvoyconfigs.cilium.io)" subsys=k8s
```

**Comment by joestringer** [comment]
@vincentmli could you file a new issue for that? It's probably something we need to fix up separately for 1.12.

**Comment by vincentmli** [comment]
sure, I will file an issue report

**Comment by vincentmli** [comment]
this is the L7 policy i applied to client pod
```
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  namespace: cilium-test
  name: client-egress-l7-http
spec:
  description: "Allow GET one.one.one.one:80/ and GET <echo>:8080/ from client2"
  endpointSelector:
    matchLabels:
      other: client
  egress:
  # Allow GET / requests towards echo pods.
  - toEndpoints:
    toPorts:
    - ports:
      - port: "8080"
        protocol: TCP
      rules:
        http:
        - method: "GET"
          path: "/"

```
then send request to VTEP from client pod

```
# kubectl exec -it client2-5998d566b4-6ssj2  -n cilium-test  -- curl -v  http://10.1.34.66:8080/
*   Trying 10.1.34.66:8080...
* Connected to 10.1.34.66 (10.1.34.66) port 8080 (#0)
> GET / HTTP/1.1
> Host: 10.1.34.66:8080
> User-Agent: curl/7.78.0
> Accept: */*
> 

```

response from VTEP:

```
< HTTP/1.1 200 OK
< date: Thu, 26 May 2022 16:06:51 GMT
< server: envoy. <=====server envoy header
< last-modified: Fri, 05 Jan 2018 23:20:34 GMT
< etag: "d75-5620fb2011480"
< accept-ranges: bytes
< content-length: 3445
< vary: Accept-Encoding
< set-cookie: Cookie=Monster; Path=/; HttpOnly
< content-type: text/html
< set-cookie: BIGipServer~AS3~A1~web_pool=1157627914.36895.0000; path=/; Httponly <====BIGIP virtual cookie
< x-envoy-upstream-service-time: 10 <=====envoy header
< 
<!--
	f5-hello-world - Main Page
	https://github.com/f5devcentral/f5-hello-world
	Artiom Lichtenstein
	v2.2.0, 19/12/2017
-->

```

**Comment by vincentmli** [comment]
denied request by envoy

```
kubectl exec -it client2-5998d566b4-6ssj2  -n cilium-test  -- curl -v  http://10.1.34.66:8080/foo
*   Trying 10.1.34.66:8080...
* Connected to 10.1.34.66 (10.1.34.66) port 8080 (#0)
> GET /foo HTTP/1.1
> Host: 10.1.34.66:8080
> User-Agent: curl/7.78.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 403 Forbidden
< content-length: 15
< content-type: text/plain
< date: Thu, 26 May 2022 16:14:27 GMT
< server: envoy
< 
Access denied
* Connection #0 to host 10.1.34.66 left intact
```

**Comment by joestringer** [comment]
:thinking: LGTM. I found this a bit confusing:

```
  # Allow GET / requests towards echo pods.
  - toEndpoints:
    toPorts:
```

Specifically the comment and the initial `toEndpoints: ` with no value doesn't seem right. But still, with the rest of your example the traffic appears to be reaching out via the proxy and potentially getting denied so it looks effective :+1:

**Comment by vincentmli** [comment]
> 🤔 LGTM. I found this a bit confusing:
> 
> ```
>   # Allow GET / requests towards echo pods.
>   - toEndpoints:
>     toPorts:
> ```
> 
> Specifically the comment and the initial `toEndpoints: ` with no value doesn't seem right. But still, with the rest of your example the traffic appears to be reaching out via the proxy and potentially getting denied so it looks effective 👍

the original policy example has  `toEndpoints:` value to another pod,  but I could not think of what value would be for VTEP, so I removed the value :)

**Comment by vincentmli** [comment]
@joestringer on a topic related to L7 proxy https://github.com/cilium/cilium/issues/19656, the client pod source ip is changed to cilium_host interface ip, I watched one of your talk on the idea of `socket_redirect`, would that solve this? any future eBPF kernel work in  socket area?

**Comment by joestringer** [comment]
It can help with that, yeah - in fact we already configure Envoy to reuse the client's IP for certain cases, but not all. If the traffic is heading out of the cluster anyway then I don't think it makes a big difference whether the cilium_host interface IP is used or the client IP, since the client IP isn't publicly routable anyway. The destination won't end up seeing that IP. Also, since the traffic is going outside the cluster, the other side won't need to directly associate that traffic back to the original pod's IP.

If you have thoughts in contrary to this where it'd be better to retain the client IP, that'd be interesting. But also be aware that reusing the client's IP is a bit more difficult to do correctly as the datapath needs to understand that this IP is actually being hijacked by the proxy, so when passing the reply traffic back, it needs to go back to the proxy rather than sending it directly to the pod with that IP.

**Comment by vincentmli** [comment]
Thanks for the answer, for now I am just curious from technical perspective and think some users might want to audit/log which source ip address accessed  server

**Comment by joestringer** [comment]
Makes sense, Envoy should be able to provide that information at L7.

**Comment by vincentmli** [comment]
Hi @pchaigno any input from you ?

File: pkg/datapath/linux/config/config.go [hunk 4]
```diff
@@ -571,6 +571,10 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 		cDefinesMap["ENABLE_ICMP_RULE"] = "1"
 	}
 
+	if option.Config.EnableL7Proxy {
+		cDefinesMap["ENABLE_L7_PROXY"] = "1"
+	}
```
Comment (pchaigno) [inline pkg/datapath/linux/config/config.go:576]: @vincentmli Any reason you added this? I don't see it used anywhere in the BPF codebase.

Comment (vincentmli) [reply pkg/datapath/linux/config/config.go:576]: I may have added it while in the middle of developing this feature and ended up not using it

Comment (pchaigno) [reply pkg/datapath/linux/config/config.go:576]: Ok. I'll send a PR to remove it.

**Pull request merged**